core/src/apresolve.rs

```diff
@@ -82,30 +97,33 @@ impl ApResolver {
         let result = self.try_apresolve().await;
 
         self.lock(|inner| {
-            let data = match result {
-                Ok(data) => data,
-                Err(e) => {
-                    warn!("Failed to resolve access points, using fallbacks: {}", e);
-                    ApResolveData::default()
-                }
+            let (data, error) = match result {
+                Ok(data) => (data, None),
+                Err(e) => (ApResolveData::default(), Some(e)),
             };
 
-            inner.data.accesspoint = self.process_data(data.accesspoint);
-            inner.data.dealer = self.process_data(data.dealer);
-            inner.data.spclient = self.process_data(data.spclient);
+            inner.data = self.parse_resolve_to_access_points(data);
+
+            if inner.data.is_any_empty() {
+                warn!("Failed to resolve all access points, using fallbacks");
+                if let Some(error) = error {
+                    warn!("Resolve access points error: {}", error);
+                }
+
+                let fallback = self.parse_resolve_to_access_points(ApResolveData::fallback());
```

Comment:
Thanks for looking into this, @dnlmlr! Unfortunately this doesn't really work as expected. The description for `--ap-port` says:

> Connect to an AP with a specified port 1 - 65535. If no AP with that port is present a fallback AP will be used. Available ports are usually 80, 443 and 4070.

But by sending the fallback data through `parse_resolve_to_access_points()` the latter would filter by the AP port again - resulting in an empty list. My understanding from the above description would be that it should ignore the requested port and fall back to whatever is hard coded.

Reply:
I think we should change the documentation. A fallback AP really just is another AP. The `ap-port` is a filter in case you require a specific port to go through your firewall, for example. If there is no AP available with the selected port, then you should not be getting any — getting one with another port would be undesired behavior (though I know it’s now been for the recent `ChannelError` issue, which this HTTP API no longer suffers from).

Reply:
That's actually a fair comment. The documentation doesn't really say whether it would use the `ap-port` value for the fallback or not.

And yes, I tripped over this because I added the invalid AP port in my application to force the use of the fallback.

Reply:
I actually bypassed the filter at first, but changed it to this again. See https://github.com/librespot-org/librespot/pull/1038#issuecomment-1202977448
The current behavior is also what it did before the PR